Refactor deck to use /prowjobs.js

katerinapat opened the pull request:
The PR is mainly refactoring of `script.js` in order to use `/prowjobs.js `. 

The only addition is the calculation of the job's duration which it is not included in the new response.

Ref https://github.com/kubernetes/test-infra/issues/5216

stevekuznetsov commented:
/ok-to-test

stevekuznetsov commented:
/cc @BenTheElder @qhuynh96

stevekuznetsov commented on prow/cmd/deck/static/script.js:
```diff
@@ -693,3 +692,18 @@ function createIcon(iconString, tooltip = "") {
+        completionMoment = new moment(completionTime),
+        d = moment.duration(completionMoment.diff(startMoment));
+
+    return d.hours() + "h" + d.minutes() + "m" + d.seconds() + "s";
```
`d.format()` ?

qhuynh96 replied:
Duration object seems not to have `format()`.

rmmh replied:
Yeah, see https://github.com/moment/moment/issues/1048.

This is wrong for jobs that take too long, and it will look weird to have a bunch of 0h jobs.

    var out = d.seconds() + 's';    
    if (d.asHours() >= 1) 
        out = Math.floor(d.asHours()) + 'h' + d.minutes() + 'm' + out;
    else if (d.minutes() >= 1) 
        out = d.minutes() + 'm' + out;
    return out;

stevekuznetsov commented on prow/cmd/deck/static/script.js:
```diff
@@ -530,9 +530,8 @@ function createTextCell(text) {
 
-// TODO FIX date time
 function createTimeCell(id, time) {
-    var momentTime = moment.unix(time);
```
Not immediately obvious why this was wrong?

stevekuznetsov commented on prow/cmd/deck/static/script.js:
```diff
@@ -26,16 +26,16 @@ function optionsForRepo(repo) {
+        if (!repo || repo === build.spec.refs.repo) {
+            opts.jobs[build.spec.job] = true;
+            if (build.spec.type === "presubmit") {
+                opts.authors[build.spec.refs.pulls[0]['author']] = true;
```
Do we want to not squash the list of pulls for batches?

qhuynh96 replied:
is there any case that build.spec.refs.pulls can be `null` or `undefined`?

stevekuznetsov replied:
Yes, in periodic and postsubmit jobs there will be no pulls

qhuynh96 commented on prow/cmd/deck/static/script.js:
```diff
@@ -418,25 +422,50 @@ function redraw(fz) {
+            build.spec.refs = prop;
+        } 
+
+        if (!build.spec.refs.hasOwnProperty("pulls")) {
```
is there any case that `build.spec.refs` inherit `pulls`? if not we can change the condition to `!build.spec.refs.pulls`

qhuynh96 commented on prow/cmd/deck/static/script.js:
```diff
@@ -418,25 +422,50 @@ function redraw(fz) {
+            ]
+        }
+ 
+        if ( Object.keys(build.spec.refs).length == 0){
```
this should be explicitly `periodic` type job as we expect `postsubmit` `presubmit` and `batch` to have `refs`; otherwise, revision cells will be filled with broken links. 
 
If `postsubmit`, `presubmit` or `batch` does not have `refs`, should we consider to drop it?

kargakis replied:
I am inclined to say just drop them but if we don't do that today I don't consider it important fixing here.

kargakis commented on prow/cmd/deck/static/.gitignore:
```diff
@@ -1,4 +1,4 @@
-data.js
```
Let's keep ignoring both until we remove data.js altogether

kargakis commented:
Looks OK overall to me. There have been some concerns raised by @rmmh and @BenTheElder in the past wrt the size of prowjobs.js. Maybe we can pass a url parameter that would drop pod specs from the response?

kargakis commented:
> Looks OK overall to me. There have been some concerns raised by @rmmh and @BenTheElder in the past wrt the size of prowjobs.js. Maybe we can pass a url parameter that would drop pod specs from the response?

Opened https://github.com/kubernetes/test-infra/pull/7386

rmmh commented on prow/cmd/deck/static/script.js:
```diff
@@ -26,16 +26,16 @@ function optionsForRepo(repo) {
+            opts.jobs[build.spec.job] = true;
+            if (build.spec.type === "presubmit") {
+                opts.authors[build.spec.refs.pulls[0]['author']] = true;
+                opts.pulls[build.spec.refs.pulls[0]['number']] = true;
```
`pulls[0].number`

rmmh commented on prow/cmd/deck/static/script.js:
```diff
@@ -353,7 +353,11 @@ function equalSelected(sel, t) {
 
 function groupKey(build) {
-    return build.repo + " " + build.number + " " + build.refs;
+    return build.spec.refs.org + "/" +
```
This won't do the right grouping for batch jobs. Try this:

    const refs = build.spec.refs;
    const pulls = refs.pulls.map(r => r.number + ':' + r.sha).join(',');
    return `${refs.repo} ${refs.base_ref}:${refs.base_sha},${pulls}`;

rmmh commented on prow/cmd/deck/static/script.js:
```diff
@@ -418,25 +422,50 @@ function redraw(fz) {
+        var build = allBuilds.items[i];
+
+        // assign default values 
+        var prop = { 
```
I would skip this, and instead change the one case that uses build.spec.refs to skip if build.spec.refs is empty.

rmmh commented on prow/cmd/deck/static/script.js:
```diff
@@ -26,16 +26,16 @@ function optionsForRepo(repo) {
         states: {},
     };
```
Why not do an initialization pass to set `build.author` and `build.number`, and avoid the construct of `build.spec.refs.pulls[0]` elsewhere?

rmmh replied:
This should still be done-- getting the hnumber out should be simpler.

kargakis commented:
@katerinapat is this ready for another round of reviews?

stevekuznetsov commented:
When we get closer to merge, please rebase your branch on top of master instead of merging.

katerinapat commented:
@kargakis. I pushed 2 more commits. Added the param `omit=pod_spec` on the request of prowjobs.js and I dropped the default values. Instead I check the cases that the values could not be set.

I'll refactor the duration so it won't display all these 0h.

 Although in this PR I didn't include any other refactoring or changing to ES6. Main goal is to refactor the script to support the new response without brake anything.  I'll be more happy to refactor the whole spaghetti code in a future PR

fejta commented:
/assign @kargakis @rmmh 
/lifecycle stale

Still working on this one?

fejta commented:
Please address comments and reopen!
/close
